Allow for excluding views from listTables()

rmarsh1ua opened the pull request:
Adds an option to listTables in Cake\Database\Schema\Collection to allow for excluding_views from a list of all tables. It seems like in many cases it's helpful to get a list of ALL tables and views together, but in the case of dropping all tables, then it's better to only work on a list of all the actual tables.

Closes #16122

rmarsh1ua committed:
Adds an option to listTables in Cake\Database\Schema\Collection
to allow for excluding_views from a list of all tables.

Closes #16122

othercorey commented:
I don't know if we need views from listTablesSql. If we did, it should probably be a separate query.

markstory commented:
> I don't know if we need views from listTablesSql. If we did, it should probably be a separate query.

We do because people use views as the datasource for ORM tables.

markstory commented on src/Database/Schema/CollectionInterface.php:
```diff
@@ -29,7 +29,7 @@ interface CollectionInterface
      * @return array<string> The list of tables in the connected database/schema.
      */
-    public function listTables(): array;
+    public function listTables(array $options = []): array;
```
This is a breaking change. We need to find a different solution to this.

rmarsh1ua replied:
Is the breaking change represented in the failing CI tests? I don't have MySQL 8 or MariaDB in my local, is there any easy way to simulate/debug changes in those environments?

This is my first time attempting to contribute back to any open source project so thank you for any guidance you can provide.

markstory replied:
>This is my first time attempting to contribute back to any open source project so thank you for any guidance you can provide.

That's great! I'm happy to help. :smile:

The breaking change is because we're adding parameters to an interface. Implementations of this interface outside of CakePHP will now raise fatal errors as the method signatures don't match. Adding a method will also fail in a similar way.

Instead what we could do is add another method to the implementations and use an `@method` comment in the class doc block. This will let IDEs know that the method exists. We'll also need to use `method_exists()` when calling the new function as static analysis tools will infer that the method doesn't exist :smile:

rmarsh1ua replied:
Okay, I think I'm picking up what you're putting down. Either today or tomorrow I'll make a new commit to the PR and you can let me know if I got it.

To my other question, some of the CI tests are failing; in one case it's just a coding standard thing, easy enough to fix. But in other cases, like the "CI / testsuite (7.4, mariadb) (pull_request)" environment/task, it's some SQL error I can't make heads or tails of. Is there an easy way to test changes against these environments so I can get in with the debugger, or in this case, does it even matter, i.e. can I just not worry about that?

Thanks so much for the help!

markstory replied:
> But in other cases, like the "CI / testsuite (7.4, mariadb) (pull_request)" environment/task, it's some SQL error I can't make heads or tails of. Is there an easy way to test changes against these environments so I can get in with the debugger, or in this case, does it even matter, i.e. can I just not worry about that?

The failures could be related to this change but that seems unlikely. If the tests continue to fail with your next changes we can look into it. As for making a local environment you can use docker to pull down the images we use in CI and run tests against them.

rmarsh1ua replied:
> 
Not totally sure I hit the mark on what you're looking for but take a look at the PR and let me know if I need to change things more.

markstory commented on src/Database/Schema/MysqlSchemaDialect.php:
```diff
@@ -38,7 +38,9 @@ class MysqlSchemaDialect extends SchemaDialect
     {
-        return ['SHOW TABLES FROM ' . $this->_driver->quoteIdentifier($config['database']), []];
+        return ['SHOW FULL TABLES FROM ' . $this->_driver->quoteIdentifier($config['database']) .
+        ((array_key_exists("exclude_views", $config) && $config['exclude_views'] == true) ? " WHERE Table_type = 'FULL TABLE'" : "")
```
Conventions for option keys are to use camelBacked style names. We would also need to include similar behavior for other database drivers as currently the support is not consistent between the various database drivers.

rmarsh1ua replied:
Please note that in the latest push I didn't change SqliteSchemaDialect because in that implementation listTablesSql already limits the results to tables.

othercorey commented:
> > I don't know if we need views from listTablesSql. If we did, it should probably be a separate query.
> 
> We do because people use views as the datasource for ORM tables.

Where do we use a table list though? When baking models?

markstory commented:
> Where do we use a table list though? When baking models?

Yes, and also during tests when we truncate tables. Some views can be truncated, and some can't. We could explore having additional methods for tests that exclude views instead of changing the existing methods.

rmarsh1ua committed:
Reverting changes to CollectionInterface method signature and
instead addding @method annotations to classes implementing the
interface. Also adding similar behavior for excluding views to
other schema dialects.
Did not make any changes to SqliteSchemaDialect because it already
excludes views from listTables results.

markstory commented:
Looking good. I'm starting to think that maybe it is bake that should change 🤔 Currently dialects are not entirely consistent on whether they include views, and excluding views and only return tables.

We could have a new method that does a `listTablesAndViews()` that bake uses. We don't typically add new methods in bugfix releases, but an exception could be made for this, or bake could temporarily not be able to generate classes for views. Having `Migrator` work with more configurations is worth going outside our normal delivery schedules.

Edit: clarify my idea more.

markstory commented:
Another potential solution that wouldn't require breaking changes would be to augment `Migrator::run()` and add option that contains a list of tables that would be omitted from the drop and truncation operations. We currently have a small amount of that already where we ignore `phinxlog` tables. Perhaps something like

```php
(new Migrator())->run(['skip' => ['custom_view', 'postgis_table']])
```

markstory commented:
Along the way I've conflated this solution which was supposed to be for `SchemaLoader` with a separate issue in the migrations `Migrator`. Because both issues came up around the same time they became blended in my thinking. The alternate solution I've proposed here would only be applicable to the migrations `Migrator`, however I think a similar approach could be used for `SchemaLoader`.

dereuromark commented:
Refs https://github.com/cakephp/migrations/pull/527 then - to symlink those.

rmarsh1ua committed:
Code formatting to comply with codings standards

rmarsh1ua committed:
Additional code formatting fixes and adding new unit tests.

rmarsh1ua committed:
Adding additional tests to verify different output from listTablesSql
depending on excludeViews config. Changed ConnectionHelper to leave
views alone as test cleanup would reproduce the original issue in
MySQL. This is because the new tests added required adding a basic
view to the test schema.

rmarsh1ua committed:
- Fixing typos and small bugs
 - Changing listTablesExcludeViews() in ConnectionHelper::truncateTables
 as well since views cannot be truncated

rmarsh1ua committed:
Attempting to fix sqlserver related bug

rmarsh1ua committed:
Fixing typo/bug

rmarsh1ua commented:
@markstory Not sure if we're still interested in this PR / set up of changes, but I've got this working to a point. I feel stuck on the oustanding failing tests. From everything I can tell there should be unit tests that cover the lines that are reported as not covered, and I'm not sure I understand the error in the coding standard check. I put the 'method' annotations in the implementations; does they also/alternatively need to be in the interface? TIA for any guidance you can provide.

dereuromark commented on src/Database/Schema/CachedCollection.php:
```diff
@@ -20,6 +20,11 @@
+ *
+ * @method array listTablesOptions(array $options = []) Get the list of tables available in the current connection.
+ *    Allows for an options array which can be used to modify the results.
+ * @method array listTablesExcludeViews() Get the list of tables, excluding any views,
```
is this maybe `array<string>` here?

dereuromark commented on src/Database/Schema/CachedCollection.php:
```diff
@@ -20,6 +20,11 @@
 /**
  * Decorates a schema collection and adds caching
+ *
+ * @method array listTablesOptions(array $options = []) Get the list of tables available in the current connection.
```
is this maybe `array<string, mixed>` here?

markstory commented:
> Not sure if we're still interested in this PR / set up of changes, but I've got this working to a point.

I have been thinking this over some more and feel that removing views from the return of `listTables()` while simpler will also have more risk. Despite support for views is already implemented inconsistently and the name `listTables` not including views some effects of excluding views where they are supported today would be:

* Bake no longer lists views as potential bake inputs.
* `FixtureManager` will no longer be able to create fixtures for views. This could impact some applications, but it shouldn't in practice as there was no way to *build* a view with fixtures previously either.
* `cake schema_cache` will no longer be able create or clear cache files for views.

I can see any of these being reported as an issue. To actually fix the problem I think we also need a new method that lists tables and views that can be used by the scenarios listed above.

I think you're on the right track with the method but I don't think we need a flexible options based system here. A method name like `listTablesAndViews()` could work. It is unlikely that we'll have more scenarios and if it comes up we can deal with it then. This will let us make changes to `listTables()` and only have it return tables like the method name implies it will do.

saeideng commented:
> This will let us make changes to listTables()

instead of changing current method
we can deprecate it

and adding new methods
 `listTablesAndViews()` or  `listFullTables()`  => all tables
 `listViewTables()` =>only views
 ` ... ` => only tables - views


changing the `listTables()` is a bit risky, not for bake / core repo , because we can replace it with new method , but for who use it as standalone package , or for who that use it directly this change could be surprise IMO , at least in bugfix and minor

rmarsh1ua commented:
> > This will let us make changes to listTables()
> 
> instead of changing current method we can deprecate it
> 
> and adding new methods `listTablesAndViews()` or `listFullTables()` => all tables `listViewTables()` =>only views `...` => only tables - views
> 
> changing the `listTables()` is a bit risky, not for bake / core repo , because we can replace it with new method , but for who use it as standalone package , or for who that use it directly this change could be surprise IMO , at least in bugfix and minor

Personally, I agree with this approach. I'm happy to continue working towards that if we're all in agreement that it's the best path forward.

If I'm about to change the code again, probably not worth looking at it now, but I'll probably still need some tips to get the code coverage tests working, and how to get the @method annotations working correctly.

Thanks!

markstory commented:
> and adding new methods
> listTablesAndViews() or listFullTables() => all tables
> listViewTables() =>only views
> ... => only tables - views

Do we need methods to get views on their own? I've only seen use cases where we need only tables (fixtures) and scenarios where we need both tables and views (schema_cache, bake) but none where we'd need only views.

A new method would be safer, but will add a methods into SchemaCollection, and Driver dialect traits as well.

rmarsh1ua commented:
I can't really think of a use case for getting a list of only views (at least nothing I would want to do programmatically). So to be clear I should:
- Change listTables back from using any kind of additional options to exclude views
- Add two functions: listTablesAndViews listTablesWithoutViews which do pretty much as they're named

If I get some thumbs up on this I'll go ahead and start working!

saeideng commented:
>Do we need methods to get views on their own?

OK, we no need to only views yet :+1:

rmarsh1ua committed:
Changing listTables to be deprecated and instead have two functions
one for getting tables without views and one for getting tables
including views. This contrasts from previous commits which basically
adjusted listTables to take an option setting to specify excluding
views.

rmarsh1ua committed:
Updating exclude views function name in sqlite unit tests

rmarsh1ua commented:
Seems like the CI checks are primarily unhappy at this point with (what I assume) are missing or incorrect method annotations. @markstory , could you take a look and let me know if this look generally okay, and maybe drop me a few hints on what's wrong?

markstory commented on src/Database/Schema/CachedCollection.php:
```diff
@@ -20,6 +20,11 @@
+ * @method array<string,mixed> listTablesAndViews() Get the list of tables available in the current connection.
+ * This will include any views in the schema.
+ * @method array<string,mixed> listTablesWithoutViews() Get the list of tables available in the current connection.
+ * This will exclude any views in the schema.
```
You shouldn't need these annotations on this class as the methods are implemented.

markstory commented on src/Database/Schema/CollectionInterface.php:
```diff
@@ -26,7 +26,7 @@ interface CollectionInterface
     /**
      * Get the list of tables available in the current connection.
-     *
+     * @deprecated
```
This interface could use the method annotations as we can't add the methods here.

markstory commented on src/Database/Schema/SchemaDialect.php:
```diff
@@ -182,13 +182,31 @@ public function dropTableSql(TableSchema $schema): array
+     *    getting tables from.
+     * @return array An array of (sql, params) to execute.
+     */
+    abstract public function listTablesWithoutViewsSql(array $config): array;
```
We will have to make these annotations as well to not break dialects in userland code. Currently they will fatal error about the missing methods.

rmarsh1ua committed:
Fixing method annotations

rmarsh1ua committed:
Fixing method annotations and cove standard fixes

rmarsh1ua committed:
Fixing annotation and additional code formatting mistakes

rmarsh1ua committed:
Fixing broken test case

rmarsh1ua committed:
Adding back real implementation for listTables for code coverage
purposes.

rmarsh1ua committed:
Fixing broken unit test

rmarsh1ua commented:
After an embarrassing number of commits, I'm still getting errors in CI because of two things: one, the listTables function being annotated as deprecated, and two, because it thinks there's new code that's not being covered by unit tests. 
I'm not sure what to do about the deprecated function because if I don't call that function then I get more code coverage errors. 
I'm not sure what to do about the current code coverage warnings because as far as I can tell, the unit tests I've added should hit the lines it's complaining about.

Any advice? @markstory 
Thanks in advance!

markstory commented on src/Database/Schema/PostgresSchemaDialect.php:
```diff
@@ -26,11 +26,45 @@
-     * @inheritDoc
+     * Generate the SQL to list the tables and views.
+     *
+     * @deprecated
```
Deprecated tags should have the version they were deprecated in as well.

markstory commented:
> I'm not sure what to do about the deprecated function because if I don't call that function then I get more code coverage errors.

You can use the `deprecated()` helper on `TestCase` to wrap logic that needs to call deprecated functions and ensure they still work and raise a deprecated warning.

>  I'm not sure what to do about the current code coverage warnings because as far as I can tell, the unit tests I've added should hit the lines it's complaining about.

I wouldn't worry too much about it. We don't collect full coverage from every database driver so there are always some gaps.

rmarsh1ua committed:
Updating deprecation annotations and fixing any calls to
newly deprecated functions

rmarsh1ua committed:
Updating version number

rmarsh1ua commented:
Okay, all the CI checks are passing except for the code coverage stuff. Take a look and let me know if there's any more changes, or what's next steps. TIA!

markstory commented on src/Database/Schema/Collection.php:
```diff
@@ -53,14 +58,51 @@ public function __construct(Connection $connection)
+     *
+     * @return array<string> The list of tables in the connected database/schema.
+     */
+    public function listTablesAndViews(): array
```
I'm wondering if we need to add this method as `listTables()` already exists and does the same thing. We could resolve the root problem by updating `ConnectionHelper` to use the new view-less methods.

rmarsh1ua commented:
My understanding is that we want to add a new function that has consistent implementation in all drivers and a more clear name so that there's no surprises, but maintain the existing function/inconsistent implementation for backwards compatibility. My suggestion would be to keep the two new functions.

markstory commented on src/Database/Schema/SqliteSchemaDialect.php:
```diff
@@ -152,9 +152,43 @@ protected function _convertColumn(string $column): array
      */
     public function listTablesSql(array $config): array
+    {
+        return $this->listTablesWithoutViewsSql($config);
```
> My understanding is that we want to add a new function that has consistent implementation in all drivers and a more clear name so that there's no surprises, but maintain the existing function/inconsistent implementation for backwards compatibility. 

I know we had discussed earlier being more strictly backwards compatible behavior. I don't think we need to be as concerned about that when the impacted driver is Sqlite. The number of impacted applications will be much lower.

rmarsh1ua replied:
I can change things to remote the listTablesAndViews related code _if_ that's what you really want. If the only advantage is not being left with a redundant/deprecated function, I'd say the work is already done and it would be better to keep it as it is now. LMK if you want me to change it though, and I'll work on that.

markstory replied:
I can take care of those changes. Thank you for putting up with all the back and forth on this and tolerating the chaos around how many and the names of the methods we add :bow:

othercorey commented on src/Database/Schema/Collection.php:
```diff
@@ -25,6 +25,11 @@
  * Used to access information about the tables,
  * and other data in a database.
+ *
+ * @method array<string> listTablesAndViews() Get the list of tables available in the current connection.
```
Why does this need to be annotated here? The function is defined below in the class.

markstory replied:
I can clean this up in a follow up pull request.

othercorey commented on src/Database/Schema/CollectionInterface.php:
```diff
@@ -21,12 +21,18 @@
+ * @method array<string> listTablesAndViews() Get the list of tables available in the current connection.
+ * This will include any views in the schema.
+ * @method array<string> listTablesWithoutViews() Get the list of tables available in the current connection.
+ * This will exclude any views in the schema.
```
If this is expected to exist only in Schema\Collection, not sure we should add an annotation on the interface. Does anything else implement CollectionInterface?

markstory commented:
Thanks for all the work you put into this @rmarsh1ua :clap:

rmarsh1ua commented:
Thanks for the guidance through this, fellas. I learned a lot in the process and I hope I'll have a chance to make some more contribs back to CakePHP in the future!
